[SPARK-30523][SQL] - Collapse nested aggregates

jaltekruse opened the pull request:
~~Note to reviewers - I'm fixing some issues after moving this onto the master branch, it was developed on the 2.4 branch and some issues weren't caught during merge resolution.~~
Ready for review, the one automated lint check failure appears to be on other PRs, and I believe is unrelated to this changeset.

### What changes were proposed in this pull request?
Combines two adjacent Aggregate operators into one, if the first one is not necessary.

If we are referencing the outputs of aggregate functions in the inner aggregate from the outer
one, check if they are being used in outer aggregates in a way that can be collapsed into a
single aggregate. A sum of sums, or a max of max, or min of min are all collapsible.
avg over avg will not be collapsible because different number of raw rows will have contributed
to the partial averages of the inner aggregate.

Min an Max can be folded in the case described above, or if they are referencing
the group by columns from the inner aggregate, as they can safely be computed just
using the set of unique values.
```
SELECT sum(sumAgg) as a, year from (
      select sum(1) as sumAgg, course, year FROM courseSales GROUP BY course, year
) group by year

// can be collapsed to
SELECT sum(1) as `a`, year from courseSales group by year
```
```
SELECT sum(agg), min(a), b from (
     select sum(1) as agg, a, b FROM testData2 GROUP BY a, b
     ) group by b

// can be collapsed to
SELECT sum(1) as `sum(agg)`, min(a) as `min(a)`, b from testData2 group by b
```

### Why are the changes needed?
Improve performance of nested aggregation queries.


### Does this PR introduce any user-facing change?
No


### How was this patch tested?
Lot's of tests added with the changeset.

jaltekruse committed:
[SPARK-30523][SQL] - Collapse nested aggregates

Combines two adjacent Aggregate operators into one, if the first one is not necessary.

If we are referencing the outputs of aggregate functions in the inner aggregate from the outer
one, check if they are being used in outer aggregates in a way that can be collapsed into a
single aggregate. A sum of sums, or a max of max, or min of min are all collapsible.
avg over avg will not be collapsible because different number of raw rows will have contributed
to the partial averages of the inner aggregate.

Min an Max can be folded in the case described above, or if they are referencing
the group by columns from the inner aggregate, as they can safely be computed just
using the set of unique values.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -964,6 +965,155 @@ object CombineFilters extends Rule[LogicalPlan] with PredicateHelper {
+                  val resolvedInnerExprs = childAggExprs.filter(
+                    ex => outerAggExpr.references.exists(_.name == ex.name))
+                  // this rule only handles cases where outer expressions are simple
+                  // and reference a single column from the inner aggregate.
```
any concern for multiple inner aggregates for now?

jaltekruse replied:
I don't believe there are lots of cases that would be collapsible, there is one case in the tests that is marked as "this should be collapsible" along these lines. Are there some in particular you are thinking of?

```
  // I think this could be collapsed, I would need to make the rule smarter to understand
  // that sums can be added together with the standard scalar expression addition operator
  test("Cannot collapse nested Agg references, complex outer agg expression with sums") {
    val query =
      """|SELECT sum(sumEarnings + sumYear), instructor from (
         |   select sum(earnings) as sumEarnings, sum(year) as sumYear, course, instructor FROM courseSalesWider GROUP BY course, instructor
         |) group by instructor
      """.stripMargin
```

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -964,6 +965,155 @@ object CombineFilters extends Rule[LogicalPlan] with PredicateHelper {
+                    val resolvedInnerExpression = resolvedInnerExprs.iterator.next()
+                    var newExpr: NamedExpression = null
+                    unwrapAlias(resolvedInnerExpression) match {
+                      case childAggExpr@AggregateExpression(_, _, _, _, _) =>
```
```scala
case childAggExpr: AggregateExpression =>
```

jaltekruse replied:
fixed

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -964,6 +965,155 @@ object CombineFilters extends Rule[LogicalPlan] with PredicateHelper {
+            case a@Alias(outerAggExpr: AggregateExpression, _) =>
+              outerAggExpr.aggregateFunction match {
+                case _: Max | _: Min | _: Sum =>
+                  // look for the expressions in the inner aggregate that produce
```
Looks like the cases can be optimized by this rule is limited:

1. Max, Min, Sum only
2. Only one inner aggregate is referred at outer aggregate
3. Outer uses subset grouping expressions of inner

Is it common to have the cases this rule can optimize?

jaltekruse replied:
It does handle the outer aggregate referring to multiple inner aggregates, there are some tests included that show this.

The rule is somewhat limited, but these cases come up, especially when it isn't a single person writing the whole query. If you are consuming a view and adding to it, you can be interested in aggregating the same types of columns that the writer of the previous view did, but your view has been rolled up to some granularity, say by region and weekly bucketing, and for your use case you want to just aggregate to straight region level.

I am happy to try to expand the cases covered by the rule to make it more useful, this is just the set of cases I was able to figure out so far.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -964,6 +965,155 @@ object CombineFilters extends Rule[LogicalPlan] with PredicateHelper {
+    plan transform {
+      // The query execution/optimization does not guarantee the expressions are evaluated in order.
+      // We only can combine them if and only if both are deterministic.
+      case agg@Aggregate(groupExprs: Seq[Expression],
```
I think you can only collapse outer and inner Aggregate if the outer has grouping expression among inner's grouping expressions?

For example:

```
SELECT sum(agg), min(a), b from (
  SELECT sum(1) as agg, max(c) as maxC, a, b FROM testData2 GROUP BY a, b
) GROUP BY maxC
```
->
```
SELECT sum(1) as `sum(agg)`, max(c) as maxC, min(a) as `min(a)`, b from testData2 GROUP BY maxC
```
?

jaltekruse replied:
I don't believe the collapse query is valid, it is referring to an aggregate expression alias in the group by, this one I think needs 2 phases of aggregates.

viirya replied:
Yea, it's not valid, but does current rule prevent it?

jaltekruse replied:
It does avoid rewriting in these cases where the outer aggregate is grouping on the results of a grouping function of the inner aggregate. This test included negative case has a similar structure to your example.

```
  test("Don't collapse necessary nested aggregates, refer to inner agg result in outer group by") {
    val query = """
                  |SELECT sum(1), agg from (
                  |   select sum(1) as agg, a, b FROM testData2 GROUP BY a, b
                  |) group by agg
                """.stripMargin
```

jaltekruse commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -25,7 +25,7 @@ import org.apache.spark.sql.catalyst.catalog.{InMemoryCatalog, SessionCatalog}
 import org.apache.spark.sql.catalyst.expressions.aggregate._
 import org.apache.spark.sql.catalyst.plans._
-import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.plans.logical.{Aggregate, _}
```
Nope, removed

jaltekruse commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -964,6 +965,155 @@ object CombineFilters extends Rule[LogicalPlan] with PredicateHelper {
+  def unwrapAlias(ex: Expression): Expression = {
+    if (ex.isInstanceOf[Alias]) ex.children.head
+    else ex
+  }
```
fixed

jaltekruse commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -964,6 +965,155 @@ object CombineFilters extends Rule[LogicalPlan] with PredicateHelper {
+      // We only can combine them if and only if both are deterministic.
+      case agg@Aggregate(groupExprs: Seq[Expression],
+                         projectionsOfAggregateNode: Seq[NamedExpression],
+                         childAgg@Aggregate(childGroupExprs, childAggExprs, grandChild)) =>
```
Fixed both, I think something in an earlier iteration of the code prevented saving these type declarations, but the current version runs fine without them.

jaltekruse commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -964,6 +965,155 @@ object CombineFilters extends Rule[LogicalPlan] with PredicateHelper {
+        //
+        // The or condition is kind of a hack for "early exit" of this loop, so later iterations
+        // don't overwrite the value to again declare this collapsible
+        if (groupExprs.exists(ex => ex.semanticEquals(aggEx)) || !collapsible) aggEx
```
Fixed

Commit:
[SPARK-30523][SQL] - Add back code to disable rule based on config property, address minor code style review comments

- While the config remained, this code block it was controlling was lost in some earlier refactoring before the
  PR was opened adding it back fixes the tests that rely on disabling the rule to determine it applies when it is supposed to

jaltekruse commented:
@viirya I believe I have addressed all of the comments. When you have some time can you take another look?

Note that along with the style fixes, I noticed that in some refactoring I did before opening the PR I had lost the small conditional that was actually implementing disabling the rule based on the new option setting. This has been added back which fixes executing the tests that rely on turning the rule on and off to ensure it is functioning properly, and all of the queries return the same results regardless if the rule is used or not.

jaltekruse commented:
@viirya do you have some time to take another look at this?

jaltekruse commented:
@viirya Do you have some time to take another look at reviewing this, or could you possibly recommend another committer that could do a review?
